### PR TITLE
JIT: Fix operand evaluation order for GT_INDEX_ADDR (#20047)

### LINKED ISSUE #20040: Incorrect behavior when TieredCompilation is enabled
Some Orleans users have hit an issue when running on .NET Core 2.1 with TieredCompilation enabled. After investigation, I believe the issue is a JIT bug related to TieredCompilation.

Repro gist: https://gist.github.com/ReubenBond/b5bdf98423d5279a8e9ae63a5e6e8f0a

In the repro, the line `return buff[offset];` will return `buff[0]` when TieredCompilation is enabled and `buff[1]` (correct) when TieredCompilation is disabled.

```C#
byte[] CheckLength(out int offset)
{
    // In the original code, this logic is more complicated.
    // It's simplified here to demonstrate the bug.
    offset = 1;
    return currentBuffer;
}

public byte ReadByte()
{
    int offset;
    var buff = CheckLength(out offset);
    return buff[offset];
}
```

We have advised the affected users to disable TieredCompilation for now, as a workaround.

### PR DESCRIPTION
**Pull request by AndyAyersMS**
We need to evaluate the array operand first, and it's op1. So evaluate
in that order, and don't allow reversal.

Closes #20040.

### CODE REVIEW

**Comment by AndyAyersMS** [comment]
Port of this to 2.2 was approved by shiproom...

cc @RussKeldorph 

See #20083 for now-abandoned port to 2.1.

**Comment by RussKeldorph** [comment]
I think x86 failures are due to a toolset update on the build machine

**Comment by AndyAyersMS** [comment]
Any ETA on resolving the toolset issue?

Aside from x86 issues, there are package restore failures and a few random disconnects.  Am going to try retesting ...

@dotnet-bot retest this please

**Comment by RussKeldorph** [comment]
@AndyAyersMS I think we may just need #20142

**Comment by RussKeldorph** [comment]
Scratch that.  Make it #20144.

**Comment by AndyAyersMS** [comment]
Retesting now that #21044 is in.

@dotnet-bot retest this please

**Comment by AndyAyersMS** [comment]
[arm64](https://ci.dot.net/job/dotnet_coreclr/job/release_2.2/job/arm64_cross_checked_windows_nt_innerloop_flow_prtest/10/) still unhappy:
```
error C3859: virtual memory range for PCH exceeded; please recompile with a command line option of '-Zm72' or greater [D:\j\workspace\arm64_cross_c---30d2bc78\bin\obj\Windows_NT.arm64.Checked\src\vm\wks\cee_wks.vcxproj
```

**Comment by AndyAyersMS** [comment]
@RussKeldorph any concerns with me merging this? I don't think those last bits of CI are reliable.

File: tests/src/JIT/Regression/JitBlue/GitHub_20040/GitHub_20040.csproj [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Debug|AnyCPU' "></PropertyGroup>
+  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release|AnyCPU' "></PropertyGroup>
+  <ItemGroup>
+    <CodeAnalysisDependentAssemblyPaths Condition=" '$(VS100COMNTOOLS)' != '' " Include="$(VS100COMNTOOLS)..\IDE\PrivateAssemblies">
+      <Visible>False</Visible>
+    </CodeAnalysisDependentAssemblyPaths>
+  </ItemGroup>
+  <PropertyGroup>
+    <DebugType>None</DebugType>
+    <Optimize>True</Optimize>
```
Comment (briansull) [inline tests/src/JIT/Regression/JitBlue/GitHub_20040/GitHub_20040.csproj]: Should; this repro be compiled as "Optimized=true"
I thought this it required MinOpts.

**Comment by AndyAyersMS** [comment]
It does require minopts, but there's no C# option set that will enable minopts for a method. 

So we either let tiered jitting or forced minopts runs cover it.

**Comment by AndyAyersMS** [comment]
Actually I stand corrected -- marking the method with `[MethodImplOptions(NoOptimization)]` should do the trick.

I will update the test.

**Pull request merged**